### ISSUE TITLE
Gallery

### ISSUE THREAD

**Issue by Andrews54757** [AJS-development/OpenAgar#1]
Revenge system:

![screen shot 2016-11-03 at 3 01 23 pm](https://cloud.githubusercontent.com/assets/13282284/19981032/98f7a28a-a1d6-11e6-814d-d9237218803a.png)
![screen shot 2016-11-03 at 3 01 29 pm](https://cloud.githubusercontent.com/assets/13282284/19981038/9c80b590-a1d6-11e6-9f8f-fb74974577f8.png)

**Comment by LegitSoulja** [comment]
Does this project even work with windows due to pipes e.g child processes?

Because it keeps freezing after 5 seconds rather or not I join the game. The only way I can get into the game is by connecting via client whilist starting the server, which will then freeze within 5 seconds again.

There was bots, and I seen that bots was still roaming around as usual, so I wouldn't consider the game frozen, but the console gets frozen where I cannot put input.

I tried looking for the issue, but I don't know exactly what it could be. You say it works for you, im guessing with Mac, but with windows, and linux it does not.

My question is, why are will still trying to use nodejs? In my opinion, nodejs is hell in a nutshell. Browser fails with javascript, I don't know what they considered usable for servers. 

I've been thinking what else could be used. Something that doesn't need as much requirements, and etc, and works for all OS. I really don't know. You'd rather use socket.io for the client, but c# doesn't support io. 

I could use regular TCP Listeners, but then how would emit specific things, and also use bytes for streaming, and sending wise. It's not going to work :/.

Also, Unity. I give up because of the browser capabilities with Unity engine. It doesn't support Chrome, and after I tried other solutions, and other browser it still does not work. Only way this will work without unity engine is if I code the client using Unity Script, which is javascript, but there type of javascript. 

So, I might as well finish with what I was doing with the old client development, which I turned mostly all objects to classes, and fixed a few things and also removed ALOT of things that wasn't needed, and removed duplicated stuff. Once I finish with just I guess the framework/model of the client ALOT of other things can be added aswell instead of everything not being organized.

Let me know if you have anything on your mind, or suggestions. I really don't have no idea what to do, and pros and cons of doing so.

**Comment by Andrews54757** [comment]
Yes, i noticed how it froze for other users. I tested it on my vm and it wouldnt work. I am going to see what I can do about that

**Comment by Andrews54757** [comment]
I think the problem is coming from socket.io. But then again, it works on the Cloud, and it is ubuntu

**Comment by Andrews54757** [comment]
When I looked at the cpu profiles, I think it maybe might be because some things update too fast, I made it so it works as seamless as possible, but it kinda uses more cpu

**Comment by LegitSoulja** [comment]
@Andrews54757 

Are you using https://github.com/socketio/socket.io-client ?

**Comment by Andrews54757** [comment]
I think I found the source. I think it might be the new Date function. It is causing garbage collector to trigger a lot and it can be fixed by just using Date.now as it reuses the variable. Im gonna see what this does

**Comment by LegitSoulja** [comment]
@Andrews54757 Scratch what I asked, that was the wrong one link.

And if you need me to test for you just let me know.

**Comment by Andrews54757** [comment]
Screws. I can't figure it out.

**Comment by Andrews54757** [comment]
I just compared the cpu to MultiOgar, And I found something interesting. I added 1000 bots for each, and played inside them. I found out that OpenAgar is a little less efficient than multiOgar, but found that this is because the collision engine is flimsy. Otherwise, Multiogar is slower. Still I noticed how the client only lagged for multiOgar, and this is with the old.ogarul.io client, but the cpu graphs say that Multi-ogar was faster. I noticed how on the graphs, that multi-ogar's collision engine is basically nonexistent. That I think, is very easily fixable for OpenAgar. I just was lazy on the collision engine. But with a little tweaks, I think I can have OpenAgar faster. One main thing I also noticed was that when Multi-Ogar lags, it doesn't stop, but instead goes a lot more slower. OpenAgar, on the other hand, lags less frequently, but freezes for a second or two. And that is the most annoying type of lag. I think I can fix that by handling some logic client-side. Currently, I would say that Multi-Ogar might be more efficient, yet that can be very easily changed, since MultiOgar is bound to other clients, and we are not. Which allows for us to do more optimizations. I am also thinking about switching over to uws, but I want to keep it simple without any buffers and whatnot.

**Comment by Andrews54757** [comment]
Data: 
[OpenAgarData.zip](https://github.com/AJS-development/OpenAgar/files/572995/OpenAgarData.zip)
[MultiOgarData.zip](https://github.com/AJS-development/OpenAgar/files/572996/MultiOgarData.zip)


PS: Good work @Barbosik, keep it up I was impressed :)

**Comment by Andrews54757** [comment]
I found the problem, it was that mainloop was synchronus. Now, I made the mainloop async. I also optimised playercollision. The server is now much more efficient than multiOgar. 
[profile-1478358866614.cpuprofile.zip](https://github.com/AJS-development/OpenAgar/files/573300/profile-1478358866614.cpuprofile.zip)


I cant believe it

**Comment by Andrews54757** [comment]
Note that there are 1000 bots running:

Before: 
![screen shot 2016-11-05 at 11 20 02 am](https://cloud.githubusercontent.com/assets/13282284/20031189/dd31efe0-a349-11e6-9b12-1cdad13f3143.png)

After: 
![screen shot 2016-11-05 at 11 19 26 am](https://cloud.githubusercontent.com/assets/13282284/20031191/e8282aa4-a349-11e6-824b-56b3e6c49918.png)

**Comment by Andrews54757** [comment]
**CURRENT EFFICIENCY AS OF NOV 5**


I made it much more efficient again. Note that 1000 bots are running. I am comparing with OgarUL


OpenAgar:

![screen shot 2016-11-05 at 7 54 37 pm](https://cloud.githubusercontent.com/assets/13282284/20034378/27c4cbfc-a392-11e6-9c97-58545d562ef3.png)


OgarUL:

![screen shot 2016-11-05 at 7 54 32 pm](https://cloud.githubusercontent.com/assets/13282284/20034379/31b574cc-a392-11e6-97bd-6e07153e26d0.png)


MultiOgar:

![screen shot 2016-11-05 at 7 59 09 pm](https://cloud.githubusercontent.com/assets/13282284/20034387/7253f9c2-a392-11e6-86c6-7107c8171f60.png)


Open agar as surpassed both multiogar and ogarUL now in preformance. OgarUl just froze when 1000 bots were added, MultiOgar slowed down and became choppy when 1000 bots were added, and OpenAgar, had no lag whatsoever which is very very surprising.

**Comment by SharkFinPro** [comment]
@Andrews54757 @LegitSoulja Can I Have A Key?
My ID: wKivqHweKapKDW1YcRkW

**Comment by Andrews54757** [comment]
[zipped.ajs.zip](https://github.com/AJS-development/OpenAgar/files/573583/zipped.ajs.zip)

**Comment by Andrews54757** [comment]
WTF: 
![screen shot 2016-11-06 at 10 38 41 pm](https://cloud.githubusercontent.com/assets/13282284/20045734/dbee05be-a471-11e6-81ba-7a904d1dbb98.png)

**Comment by Andrews54757** [comment]
![screen shot 2016-11-06 at 10 42 02 pm](https://cloud.githubusercontent.com/assets/13282284/20045806/47c51318-a472-11e6-8d65-6e55a20a6924.png)

**Comment by LegitSoulja** [comment]
@Andrews54757 Not again -_-

**Comment by Andrews54757** [comment]
SEE THIS: 
![screen shot 2016-11-06 at 10 43 30 pm](https://cloud.githubusercontent.com/assets/13282284/20045866/74ad1cea-a472-11e6-96d9-ae1e674a5e83.png)

**Comment by Andrews54757** [comment]
IDIOT!!

**Comment by agariotips** [comment]
You're fucking kidding me not that trydox kid

**Comment by Andrews54757** [comment]
I resolved it. I retracted his usage of ogarUL remotly. His servers are now down

**Comment by agariotips** [comment]
Good. He's stealing @MastaCoder 's login system. Or atleast trying to. I don't understand why

**Comment by LegitSoulja** [comment]
@agariotips this kid gonna get taken down. I've already took his old site down -> play.trydox.tk

**Comment by Andrews54757** [comment]
Oh, my god, he recreated his server after I deleted them!. Oh well, if you strike a tree many times, it will eventually fall

**Comment by LegitSoulja** [comment]
@Andrews54757 I've made a program to check all his open ports. Do you want me to re-run it so you can get the stat ports?

I should make a system that automatically detects his ports, and send you his uid's lol.

**Comment by Andrews54757** [comment]
He is online and I just shut him down again

**Comment by Andrews54757** [comment]
And he now put "special thanks to ajs development" in the bottom right now of the client

**Comment by LegitSoulja** [comment]
@Andrews54757 lol, His host is this btw http://www.bplaced.net

**Comment by LegitSoulja** [comment]
I'm still waiting on a reply from there custom services

**Comment by Andrews54757** [comment]
I just gave this link to him. Expect a reply, or I will PERMENATLY delete his OgarUL

**Comment by LegitSoulja** [comment]
Im going to get his site files, and figure out if he has any php files that contains sensitive information so I can delete everything on his site.

Should be pretty easy changing the header to read the php file. His site has no security whatsoever.

**Comment by Andrews54757** [comment]
Not yet, I want to see his reasoning. He will reply here

**Comment by LegitSoulja** [comment]
@Andrews54757 ok, and one more thing. Tell me how the setuint works with the nodes so that I can scramble them.

Or ill pass you the client, and pass it back to me and ill upload it

**Comment by Andrews54757** [comment]
Wow, he just changed his server to another from the banned one. Welp, I did it again to him. And @LegitSoulja I give you permission to take action against him. And I dont even know how setuint really works

**Comment by LegitSoulja** [comment]
@Andrews54757 no luck with getting access into his site FTP. I couldn't find anything valuable.

I'll have to keep digging till I get some type of information.

**Comment by Andrews54757** [comment]
I just rolled out an update that 

1. Fixes chat commands: This was really annoying for me
2. Prepares for anti-minion updates

**Comment by agariotips** [comment]
You need to rerun his ports. They're back.

**Comment by agariotips** [comment]
You know what? I'm tired of him stealing it. I'm going to make a client for him, so he doesn't do it again.

**Comment by agariotips** [comment]
![steal](https://cloud.githubusercontent.com/assets/21321103/20121555/37031466-a5e0-11e6-9dd5-27e19d4c95ac.PNG)

**Comment by LegitSoulja** [comment]
@agariotips I warned him yesterday about me dealing lawfully with him. He made the right decision and closed his site down. Hopefully, he will do things the right way instead of stealing other people codes.

**Comment by LegitSoulja** [comment]
@Andrews54757 check gitter please. Thanks.

**Comment by Andrews54757** [comment]
Done

**Comment by agariotips** [comment]
Yeah he said he won't do anything that will make Andrew mad or you mad. I hope i can believe him.

**Comment by Andrews54757** [comment]
Tell him that the only thing we don't like his him copying our skins as well as our skin selector tool. Thats all, if he uses the client available to the public that I provided, then it is okay with me

**Comment by Andrews54757** [comment]
https://github.com/AJS-development/client/tree/gh-pages -> public client

**Comment by MastaCoder** [comment]
@Andrews54757 That's interesting. Does it contain all the features of MultiOgar yet? I meant as in, is a slim model of the Ogar version with just FFA and physics?

**Comment by Andrews54757** [comment]
Almost all the basics, except viruses. Yet the infrastructure is complete. Every feature that I add has virtually no performance drawbacks as they are made with extreme caution. More is being added. So these stats are probably off by a bit. But then, as I add more features, I also make it more efficient. So it balances out. I keep my CPU profiler at hand all the time to monitor the server

**Comment by Andrews54757** [comment]
There are also cool features as well. Such as the revenge icon. Most features however, are devoted to efficiency. There are soo many features for efficiency, It is hard to list them all. Oh well, Il try


1. Server side caches
2. Client side caches
3. Slow FPS protocol
4. Hashes
5. Hashbounds
6. SmartNodes protocol 
7. Lag detection. (the server will adapt to lag)
8. Async console
9. and more I can't remember

**Comment by MastaCoder** [comment]
Interesting 👍

**Comment by Andrews54757** [comment]
I find it depressing that Ogar is dead. No ogar server was ever updated in the last 2 months. Agar.io is apparently dead.

**Comment by LegitSoulja** [comment]
@Andrews54757 That's why we need to bring it back to life, not only as it being similar to Agar.io, but as to what you created Unlimited. Private servers still exist, not updates nah. That's hardly never heard from miniclip, except them patching and making everything look better than before.

**Comment by omidh28** [comment]
@Andrews54757 I don't think it's dead yet, Personally I'm looking for hosting hundreds of mini Ogar servers.

**Comment by MastaCoder** [comment]
@LegitSoulja The whole game concept is dead now. The game used to be fun, but games are only fun for a bit. Unless if there would be a gamemode that would completely blow out all other .io games than it would work. But the issue is now there are hundreds of .io games that just have a MUCH better concept than Agar.io, after all, Agar.io was the first .io hit.

**Comment by Andrews54757** [comment]
Yes, but I am thinking of adding special things such as mini games

**Comment by DaAwesomeRazor** [comment]
@Andrews54757 Thats amazing, my multiogar server lags a ton in the bots gamemode when i give people more than 9 bots each,

**Comment by Megabyte918** [comment]
I'm so hype for this project!

**Comment by Andrews54757** [comment]
New performance charts: This compares the performance before and after [this](https://github.com/AJS-development/OpenAgar/commit/ec767b76eb87d2bc506673768532e7449b847072)


# BEFORE:

![screen shot 2016-11-23 at 9 53 06 am](https://cloud.githubusercontent.com/assets/13282284/20566380/deba8936-b162-11e6-92d4-3f8b643c9d76.png)

# AFTER

![screen shot 2016-11-23 at 9 53 11 am](https://cloud.githubusercontent.com/assets/13282284/20566384/e745e230-b162-11e6-8414-fe00f7cedc6d.png)

**Comment by Andrews54757** [comment]
I tried 2000 bots, and there was NO LAG!!

**Comment by Andrews54757** [comment]
Well, next up: 3000 bots. And smart bots too.

**Comment by LegitSoulja** [comment]
@Andrews54757 Quick question about using a lot of if statements for no reason.

For example, which is faster for processing

This
`
function Exec(exotic = 1, player = 1)
{
    document.write(exotic * player)
}
Exec(5,5);
`

Or 

`
fuction Umm(player)
{
    if(!player) player = 1;
    document.write(player)
}
Umm();
`

**Comment by Andrews54757** [comment]
Wait, but the above is multiplying, the bottom, is just setting

**Comment by Andrews54757** [comment]
Is this on purpose

**Comment by LegitSoulja** [comment]
@Andrews54757 ignore it, it's just an example of it she use

**Comment by LegitSoulja** [comment]
Focus on the function itself and using IF statement she or not to check for null values. Why not just set a value instead of checking to see if it has no value?

**Comment by Andrews54757** [comment]
I don't know. Im dealing with polars bots right now.

**Comment by LegitSoulja** [comment]
/\ that's the least of the worries if you want things more efficient @Andrews54757 and I apologize for my grammar, auto correct is a pain on mobile.

**Comment by Andrews54757** [comment]
@LegitSoulja They are trying to get by our protection. I am talking with them to ruin them

**Comment by LegitSoulja** [comment]
@Andrews54757 don't worry. They said 2 days ago they bypassed it. Don't waste time on them kids until they really do bypass it.

**Comment by LegitSoulja** [comment]
Anyways igtg.

**Comment by LegitSoulja** [comment]
They can't bypass it. That p4rtyguy has weak knowledge btw.

**Comment by Andrews54757** [comment]
Anyway, they are ruined. I ruined them, so they are desperate to retaliate

**Comment by LegitSoulja** [comment]
I got more solutions to buff up security aswell. I'll implement them when I get back home. Currently in Cali right now.

**Comment by LegitSoulja** [comment]
@Andrews54757 read that last post on the forums under Anti-Script detection to get the idea.

**Comment by Andrews54757** [comment]
@LegitSoulja They are so mad right now. I pissed them off.

**Comment by LegitSoulja** [comment]
Hahaha. Let them be.

**Comment by Andrews54757** [comment]
They keep banning me, but they don't know what DHCP is, so I keep coming back.

**Comment by Andrews54757** [comment]
@LegitSoulja Set up a monitoring system for statistics. So we can truly know if they bypassed.


So have it detect a spike of players in the server

**Comment by LegitSoulja** [comment]
@Andrews54757 a monitoring system for statistics? The plugin? I can't because I no longer have the source.

And I'm not doing anything today cause I'm not at home. I stated earlier I was in Cali. I only have a phone with me.

**Comment by LegitSoulja** [comment]
I was working on statistics one day and I guess the domain had issues and statistics forced updated with the encrypted plugin version, so basically I lost the work :/

**Comment by Andrews54757** [comment]
@LegitSoulja , I meant from our servers. Using php, and mysql, we can monitor servers

**Comment by LegitSoulja** [comment]
@Andrews54757 be my guest. You can use FTP. Live/data/stats/

**Comment by LegitSoulja** [comment]
@Andrews54757 I downloaded gitter on my phone. We can chat there if you like

**Comment by Megabyte918** [comment]
ayy lmao

**Comment by Andrews54757** [comment]
@LegitSoulja OpenAgar is basically finished. There are only two things I have to do

1. Smarter bots. Im planning on making the bots extremely smart (very easy)
2. More gamemodes

**Comment by LegitSoulja** [comment]
@Andrews54757 I'll check out making new gamemodes when I get home.

I don't know when exactly. I might not go back home till this weekend :/.

**Comment by LegitSoulja** [comment]
@Andrews54757 Check gitter

**Comment by Niall7459** [comment]
This looks very nice! great work!

**Comment by Megabyte918** [comment]
@Andrews54757 I have a question, what was the starting mass of the bots when you ran them? I'd like to compare the performance to my version of MultiOgar, but I need to know the correct settings first.

**Comment by Andrews54757** [comment]
Default

**Comment by Andrews54757** [comment]
It takes a lot for OpenAgar to lag. Note that bots run on a different process, the same process in which collisions, leaderboard, and mass checking runs in. So they might be slowed down when bots are flooded. But the main physics wont. This only occurs for bots (not minions or players) because the bots run on the same process, and they are not optimised at all. Still though, you can run 2000 bots with no lag (for me), without the effect I mentioned. But I bet if the bot code was optimised (it is the same AI as OgarUL), it will be able to run 2500 bots without the effect. But still, lag is not visible even if you add more bots than expected (for me). The server is made to detect lag and mitigate it, you can see some lag status by going into the debug console (Type `debug`).

**Comment by LegitSoulja** [comment]
@Andrews54757 Does running a child process do it happen on the same core, or all?

In that case I got a 8 core, 64 gb mem, OS running on an SSD with 2 more internal, and processing speed is 5.7 (Overclocked). I flashed my motherboard :) in case you're wondering. That's how custom built my computer is.

**Comment by Andrews54757** [comment]
So, after studying nodejs processes, I learned a v8 engine is very expensive. And using .fork spawns a new instance. So I have it set up dynamically to get the most bang out of my buck.


1. A new child will spawn when a server is created. Including using the multi-server system. 
2. If the amount of processes, equals the amount of cpus, it will not create another but will assign the server to the process with the least assigned servers.
3. When a server is removed, (not shut down), it will deassign the server from its process
4. If the number of assigned servers go to 0, the process will be shut down

**Comment by LegitSoulja** [comment]
@Andrews54757 Good note.

**Comment by Andrews54757** [comment]
I am kindof dissapointed though, that people are trying to steal my work, I just realised, the entire advancedness (a word?) of the program and how much work it took me to make

**Comment by Andrews54757** [comment]
I just posted an update a while ago that changes the UID system so the keygen is different now

**Comment by Andrews54757** [comment]
Now I realise that I should have kept this project completely private

**Comment by Andrews54757** [comment]
Oh well, this will be good on my resume anyway

**Comment by Andrews54757** [comment]
@LegitSoulja Please go to gitter

**Comment by LegitSoulja** [comment]
@Andrews54757 Idk why you had it on github. I thought it was going to be private aswell, but it didn't turn out that way.

**Comment by Megabyte918** [comment]
@Andrews54757 I know you are probably busy and all and its the holidays and all that, but I would like to ask a few more questions.

How is the performance with 5000 ten mass minions? Be brutally honest, I'd like to compare performance again. If you do test it, can you leave the contents of the `debug` command here? Did the server slow down or physics become choppy? Please explain in great detail, I want to feel as if I tested it for myself.

I know I've already asked a similar question and you probably don't have time to answer, but I couldn't wait any longer and the forums are closed as of now.  Since I missed my chance to generate my own key, I cannot test the performance myself. I figured I would ask someone who has access to it.

Thank you for your time :)

**Comment by LegitSoulja** [comment]
@Megabyte918 Why should it matter?. Aren't you working with MastaCoder? 

By the looks of your `accomplishments`, is it even worth being involved with?

**Comment by Andrews54757** [comment]
@Megabyte918 This is what will happen with 5000 minions: The server will lag badly.


```
[Debug] Client (ID: 1) was added to server 1                                   
[Debug] Possible lag spike detected: 187 MS. Nodecount: 5671                   
[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 137 MS. Nodecount: 5671                   
[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 127 MS. Nodecount: 5671                   U[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 144 MS. Nodecount: 5671                   
[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 87 MS. Nodecount: 5671                    
[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 94 MS. Nodecount: 5671                    
[Debug] Mitigating lag...                                                      
[Debug] Possible lag spike detected: 91 MS. Nodecount: 5671                    
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 97 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 101 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 146 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 131 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 91 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 90 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 134 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 134 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 109 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 95 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 113 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 112 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 81 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 115 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 91 MS. Nodecount: 5671
[Debug] Mitigating lag...
[Debug] Possible lag spike detected: 95 MS. Nodecount: 5671
[Debug] Mitigating lag...
```

**Comment by Andrews54757** [comment]
@Megabyte918 I will however, try to make preformance even better so it can handle 5000 minions. The reason for the horrible lag at 5000 minions is because they don't run on separate processes, unlike bots.

**Comment by Andrews54757** [comment]
@Megabyte918 I have experimented more, and found that the limit for no lag minions are 2000

**Comment by Andrews54757** [comment]
@Megabyte918 I have some ideas on how to optimize them.

1. When minions are not in view of players (not bots), don't do collision checking.
2. When two minions overlap, do collision checking only once for them
3. More...

**Comment by Andrews54757** [comment]
@Megabyte918 I think, however, the lag is caused when huge amount of data is being sent to child processes. I could use a database to store the objects, and that could solve the problem, but its the same thing Micheal and I tried to do in the dev branch of ogarul. Didn't end up too well as people were complaining about the slow start time.

**Comment by Andrews54757** [comment]
@Megabyte918 


Normal Operation:

![screen shot 2016-12-23 at 8 26 55 am](https://cloud.githubusercontent.com/assets/13282284/21454842/af674824-c8e9-11e6-9479-101cfe8832b7.png)


2000 minions:

![screen shot 2016-12-23 at 8 26 34 am](https://cloud.githubusercontent.com/assets/13282284/21454846/b9834286-c8e9-11e6-9517-54055c1b321b.png)

5000 minions:

![screen shot 2016-12-23 at 8 26 31 am](https://cloud.githubusercontent.com/assets/13282284/21454850/c138816c-c8e9-11e6-910d-666fff44d046.png)

**Comment by Andrews54757** [comment]
@Megabyte918 I also learned, that a lot of the lag I experienced, was infact client lag, even though I shut off the server, it still was "moving".

**Comment by Megabyte918** [comment]
@Andrews54757 Very very interesting... I have a few things to state about the possible optimizations:
> When minions are not in view of players (not bots), don't do collision checking unless the minions are set to collect food or if their count is below 1000.

Won't that cause bugs with collision for other players trying to eat the minions? For example if I ran 2000 minions, and other players were trying to eat them out of my viewbox, if the collision checking is disabled, then the players won't be able to eat/feed/interact with the minions right?

> When two minions overlap, do collision checking only once for them

Wouldn't having the server constantly checking if minions are overlapping decrease performance? Especially if your running greater than the recommended amount of minions

**Comment by Megabyte918** [comment]
@Andrews54757 Have you tried disabling visible-node calculations for minions? It really helps. Here is an example of it in action on my server:

Disabled(default): https://streamable.com/6om8h
Enabled: https://streamable.com/xx9vz

I am using an extension because my computer can't handle recording with 5k minions with the agario client's graphics.

That being said I am very very impressed on how well OpenAgar handles 2k minions, especially with all features it has and all of the configs and all that. I'm glad it didn't turn out like OgarUL did with the lag and all, you did a great job Andrew. Thanks for taking the time to respond to all of my questions with such detail, I wish you guys good luck and happy holidays :)

**Comment by Andrews54757** [comment]
@Megabyte918 Minions already don't have visible calculations and whatnot. They are very very simplified. Take a look at the code and you will see the pure simplicity.

**Comment by Megabyte918** [comment]
@Andrews54757 Nice, I can't wait to see what it can do after your future optimizations :D

**Comment by Andrews54757** [comment]
@Megabyte918 I have discovered something: Socket.io is very slow. I found out that OpenAgar is not lagging minions due to the fact that socket.io is slow. I then pried through Socket.io's code, and made if much more efficient.

**Comment by Megabyte918** [comment]
@Andrews54757 Wow.. very nice 👍

**Comment by KaanAga** [comment]
Hey Andrews54757 I dont know "How to Setup OpenAgar?" Plz record video and send link plsss I want this OpenAgar

**Comment by Andrews54757** [comment]
@LegitSoulja, should we make the packets binary? (that didnt make so much sense). I created a tool to do so: https://github.com/ThreeLetters/Binary

**Comment by Andrews54757** [comment]
@LegitSoulja I am creating a more detailed documentation here: https://ajs-development.github.io/OpenAgar/

**Comment by LegitSoulja** [comment]
@Andrews54757 If it turns out with better performance then sure

**Comment by LegitSoulja** [comment]
I kindof want to make my own socket in which with non nodejs, but with browser, in which can communicate with socket.io, and websockets using protocols.

Or even my own sys, in which sends vice versa data.

**Comment by Andrews54757** [comment]
@LegitSoulja Wow, there are 39 members already! And a bunch of errors were also logged so I fixed them.

**Comment by Andrews54757** [comment]
@LegitSoulja There is a problem with the OAStats server section of the website, please go on gitter

**Comment by LegitSoulja** [comment]
There's 43 now lol

**Comment by LegitSoulja** [comment]
@FlamePlayerTR 

I have told people of the community not to make videos about setting up OpenAgar. OpenAgar only for those people who understands, and have knowledge of using this.

IF you have used multiogar, Ogar, or any other related Agar.io private servers it's all the same.

**Comment by LegitSoulja** [comment]
However, no more further answers, or questions should be asked. Read the readme. Case closed.

**Comment by FlamePlayerTR** [comment]
@LegitSoulja 
It was before start.bat 
But not now
Previously setup node to Ogar Folder but now Keys , panels and not start.bat
and Is this system just for mac?

**Comment by Andrews54757** [comment]
No, I can start this on windows too. Start.bat was a script for noobs. Here, we dont accept them

**Comment by LegitSoulja** [comment]
Better learn how to use cmd @FlamePlayerTR

**Comment by FlamePlayerTR** [comment]
@Andrews54757 and a video plz ?

**Comment by LegitSoulja** [comment]
It's not gonna happen.

**Comment by LegitSoulja** [comment]
<details>
  <summary>Spoilers/Noob</summary>
  I've seen some people, like valkyy, and seif who took the batch files from ogarul lol












![](https://cloud.githubusercontent.com/assets/379606/17268148/f8478fc2-55d5-11e6-929a-06ce77fef04b.gif)
</details>

**Comment by Megabyte918** [comment]
lmao what was that

**Comment by Andrews54757** [comment]
@Megabyte918 I am making sending nodes use buffers. See #50. I tested 9000 minions. The server lasted for 30 sec with no lag with that amount

![screen shot 2017-01-06 at 9 37 28 pm](https://cloud.githubusercontent.com/assets/13282284/21738707/890c4a06-d458-11e6-8da1-9209934ce9c1.png)

**Comment by Andrews54757** [comment]
@Megabyte918 , the charts look the same, performance wise, however though, the network can handle it this time. 

![screen shot 2017-01-06 at 9 43 39 pm](https://cloud.githubusercontent.com/assets/13282284/21738752/8351b0e6-d459-11e6-9828-bb5f8ecfb7c6.png)

**Comment by Andrews54757** [comment]
@Megabyte918 As you can see, the bottleneck is the updatehash function. It is probably due to a error I made in hashbounds. In the future, when I solve this, I would probably be able to run 13000 minions

**Comment by Andrews54757** [comment]
@Megabyte918 I reduced the total time UpdateHash takes by 300 ms. I am going to find more optimisations as I go

**Comment by Megabyte918** [comment]
@Andrews54757 Hmm,..very interesting, maybe I should implement hashing too

**Comment by charliebrown164** [comment]
after the long fight about the video tutorial i see it happened
by the way i appreciate what you guys are doing
i'm learning things thanx to you
i'll be sending feedback often

**Comment by LegitSoulja** [comment]
@charliebrown164 I decided why not. Besides, Andrew wants people to run OpenAgar. I seen that nobody bothered in trying, so I gave a simple tutorial just to get you started with the setup.. Mainly the key installation, and making your server public.

There's more things you'll have to figure out on your own.

**Comment by charliebrown164** [comment]
blackholes make no sense
if you are split to 16 pieces and one of the small ones come in contact with it it will teleport you to some other hole with just the mass of the little piece ?
i mean to me it's just an annoying black circle that i try to avoid

**Comment by LegitSoulja** [comment]
@charliebrown164 and it's annoying enough, that's there's an option to reduce black holes. It's not part of a gamemode, it's an actual feature. (I think).

You can change this in physicsConfig `minWormHole`

**Comment by LegitSoulja** [comment]
Also @Andrews54757 there was a bug in which kindof ruined the video lol. It just spammed numbers in the console. Couldn't provide tuts on using commands, minions, bots and such. I was already recording when I witnessed it.

**Comment by charliebrown164** [comment]
cool
no more spammed numbers
when i use the command update does it update all or i still need to download files from github like the physicsConfig.ini that was updatet 30 minutes ago but still the same on my files

**Comment by Andrews54757** [comment]
Configurations are not updated

**Comment by charliebrown164** [comment]
one more thing please
if you noticed, cells get eaten by larger cells as soon as the come in contact with each other
how can i change it so that the bigger one actually overlaps the other, to make the game fair

**Comment by charliebrown164** [comment]
is it here ? OpenAgar-master/ source/ entities/ template.js (line 186)

getEatRange() {

        return this.size * 0.4;

    }
    collisionCheckCircle(node, no) {

        var distance = this.getDistance(node.position.x, node.position.y);

        var s = (no) ? node.size : node.getEatRange(node)
        if (distance <= this.size + s) return true;
        return false;

I also want to modify it for mothercells, it's so weired to get absorbed  as soon as you touch an opponent or a mothercell
